lib/usagereporter/usagereporter.go

```diff
@@ -274,49 +261,82 @@ func (r *UsageReporter[T]) Run(ctx context.Context) {
 	r.Debug("usage reporter is ready")
 
 	for {
+		var subQueue chan []*SubmittedEvent[T]
+		var subBuf, subRest []*SubmittedEvent[T]
+		if len(r.buf) >= minBatchSize {
+			subQueue = r.submissionQueue
+			subBuf, subRest = bufRest()
+		}
+
 		select {
 		case <-ctx.Done():
+			if len(r.buf) > 0 {
+				r.WithField("discarded_count", len(r.buf)).Warn("dropped events due to context close")
+			}
 			return
+
 		case <-r.eventsClosed:
-			r.enqueueBatch()
+			for len(r.buf) > 0 {
+				subBuf, subRest := bufRest()
+				select {
+				case <-ctx.Done():
+					r.WithField("discarded_count", len(r.buf)).Warn("dropped events due to context close during graceful stop")
+					return
+				case r.submissionQueue <- subBuf:
+					usageBatchesTotal.Inc()
+					r.WithField("batch_size", len(subBuf)).Debug("enqueued batch of usage events during graceful stop")
+					r.buf = subRest
+				}
+			}
 			return
-		case <-timer.Chan():
-			// Once the timer triggers, send any non-empty batch.
-			timer.Reset(r.maxBatchAge)
-			r.enqueueBatch()
-		case events := <-r.events:
-			// If the buffer's already full, just warn and discard.
-			if len(r.buf) >= r.maxBufferSize {
-				r.WithField("discarded_count", len(events)).Warn("usage event buffer is full, events will be discarded")
 
-				usageEventsDropped.Add(float64(len(events)))
-				break
+		case <-timer.Chan():
+			minBatchSize = 1
+
+		case subQueue <- subBuf:
+			usageBatchesTotal.Inc()
+			r.WithField("batch_size", len(subBuf)).Debug("enqueued batch of usage events")
+			r.buf = subRest
+			minBatchSize = r.minBatchSize
+
+			if !timer.Stop() {
+				select {
+				case <-timer.Chan():
+				default:
+				}
+			}
+			if len(r.buf) > 0 {
+				timer.Reset(r.maxBatchAge)
 			}
 
+		case events := <-r.events:
 			if len(r.buf)+len(events) > r.maxBufferSize {
 				keep := r.maxBufferSize - len(r.buf)
+				if keep < 0 {
+					keep = 0
+				}
+
 				r.WithField("discarded_count", len(events)-keep).Warn("usage event buffer is full, events will be discarded")
 				events = events[:keep]
 
 				usageEventsDropped.Add(float64(len(events) - keep))
 			}
 
+			if len(events) <= 0 {
+				break
+			}
+
+			// about to become nonempty
+			if len(r.buf) == 0 {
+				timer.Reset(r.maxBatchAge)
```

Comment:
I'm trying to understand why we need to reset the timer here and in `case subQueue <- subBuf`. Why can't we just do it without any condition in `subQueue <- subBuf` case?

Reply:
If we completely drain the buffer the timer shouldn't run, and it should start anew only from the first new event, otherwise we risk firing off batches due to "age" with a single event in them just because it's been a while since any event has been emitted.